sdk/compatability/src/env.rs

```diff
@@ -0,0 +1,456 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::{format_err, Result};
+use diem_sdk::{
+    client::{AccountAddress, BlockingClient, SignedTransaction, WaitForTransactionError},
+    crypto::{ed25519::Ed25519PrivateKey, ValidCryptoMaterialStringExt},
+    move_types::language_storage::TypeTag,
+    transaction_builder,
+    types::{
+        account_config::{xdx_type_tag, xus_tag},
+        chain_id::ChainId,
+        transaction::{authenticator::AuthenticationKey, RawTransaction, TransactionPayload},
+    },
+};
+use once_cell::sync::OnceCell;
+use std::env;
+
+const JSON_RPC_URL: &str = "JSON_RPC_URL";
+const FAUCET_URL: &str = "FAUCET_URL";
+const MINT_KEY: &str = "MINT_KEY";
+
+// const TESTNET_JSON_RPC_URL: &str = "http://testnet.diem.com/v1";
+// const TESTNET_FAUCET_URL: &str = "http://testnet.diem.com/mint";
+
+pub struct Environment {
+    json_rpc_url: String,
+    coffer: Coffer,
+    chain_id: ChainId,
+}
+
+impl Environment {
+    pub fn from_env() -> &'static Self {
+        static ENV: OnceCell<Environment> = OnceCell::new();
+
+        //TODO don't use testnet
+        // env::set_var(JSON_RPC_URL, TESTNET_JSON_RPC_URL);
+        // env::set_var(FAUCET_URL, TESTNET_FAUCET_URL);
+
+        ENV.get_or_init(|| Self::read_env().unwrap())
+    }
+
+    fn read_env() -> Result<Self> {
+        let json_rpc_url = env::var(JSON_RPC_URL)?;
+        let coffer = if let Ok(s) = env::var(MINT_KEY) {
+            Ed25519PrivateKey::from_encoded_string(&s).map(Coffer::MintKey)?
+        } else {
+            env::var(FAUCET_URL)
+                .map(|f| FaucetClient::new(f, json_rpc_url.clone()))
+                .map(Coffer::Faucet)?
+        };
+
+        let chain_id = ChainId::new(
+            BlockingClient::new(&json_rpc_url)
+                .get_metadata()?
+                .into_inner()
+                .chain_id,
+        );
+
+        Ok(Self::new(json_rpc_url, coffer, chain_id))
+    }
+
+    fn new(json_rpc_url: String, coffer: Coffer, chain_id: ChainId) -> Self {
+        Self {
+            json_rpc_url,
+            coffer,
+            chain_id,
+        }
+    }
+
+    pub fn json_rpc_url(&self) -> &str {
+        &self.json_rpc_url
+    }
+
+    pub fn client(&self) -> BlockingClient {
+        BlockingClient::new(self.json_rpc_url())
+    }
+
+    pub fn coffer(&self) -> &Coffer {
+        &self.coffer
+    }
+
+    pub fn chain_id(&self) -> ChainId {
+        self.chain_id
+    }
+
+    pub fn transaction_factory(&self) -> TransactionFactory {
+        TransactionFactory::new(self.chain_id())
+    }
+}
+
+pub enum Coffer {
+    Faucet(FaucetClient),
+    MintKey(Ed25519PrivateKey),
+}
+
+impl Coffer {
+    pub fn fund(&self, currency: Currency, auth_key: AuthenticationKey, amount: u64) -> Result<()> {
+        match self {
+            Coffer::Faucet(faucet) => faucet.fund(currency, auth_key, amount),
+            Coffer::MintKey(_) => todo!(),
+        }
+    }
+}
+
+pub struct FaucetClient {
```

Comment:
I think faucet client maybe required for various use cases. we may consider pull it into separate package for sharing.

Reply:
Agreed, if you noticed this file tends to have a few things that probably need to be moved out so they're more reusable.